File: Data Structures & Algorithms/count-number-of-islands/submission-3.cpp
class Solution {
    private:
        int rows;
        int cols;

        void dfs(int row, int col, vector<vector<char>>& grid) {
            if (row < 0 || col < 0 || row >= rows || col >= cols || grid[row][col] != '1') {
                return;
            }

            grid[row][col] = '#';
            dfs(row + 1, col, grid);
            dfs(row - 1, col, grid);
            dfs(row, col + 1, grid);
            dfs(row, col - 1, grid);
        }

    public:
        int numIslands(vector<vector<char>>& grid) {
            if (grid.empty()) {
                return 0;
            }

            int num_of_islands = 0;
            rows = grid.size();
            cols = grid[0].size();

            for (int row {0}; row < rows; ++row) {
                for (int col {0}; col < cols; ++col) {
                    if (grid[row][col] == '1') {
                        dfs(row, col, grid);
                        num_of_islands++;
                    }
                }
            }

            return num_of_islands;
        }
};

File: Data Structures & Algorithms/surrounded-regions/submission-3.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board or not board[0]:
            return

        ROWS = len(board)
        COLS = len(board[0])

        def capture(row, col):
            stack = []
            stack.append((row, col))

            while stack:
                curr_row, curr_col = stack.pop()
                if (curr_row < 0 or curr_col < 0 or curr_row >= ROWS or curr_col >= COLS or board[curr_row][curr_col] != 'O'):
                    continue

                board[curr_row][curr_col] = '#'
                stack.extend(((curr_row + 1, curr_col), 
                              (curr_row - 1, curr_col), 
                              (curr_row, curr_col + 1), 
                              (curr_row, curr_col - 1)))

        for row in range(ROWS):
            for col in (0, COLS - 1):
                if board[row][col] == 'O':
                    capture(row, col)

        for col in range(COLS):
            for row in (0, ROWS - 1):
                if board[row][col] == 'O':
                    capture(row, col)

        for row in range(ROWS):
            for col in range(COLS):
                if board[row][col] == 'O':
                    board[row][col] = 'X'
                elif board[row][col] == '#':
                    board[row][col] = 'O'

        